File: lintcode/remove_nth_node_from_end.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of linked list.
    @param n: An integer.
    @return: The head of linked list.
    """
    def removeNthFromEnd(self, head, n):
        nth = head
        for _ in xrange(n):
            nth = nth.next
        if not nth:
            return head.next
        current = head
        while nth.next:
            nth = nth.next
            current = current.next
        current.next = current.next.next
        return head

File: leetcode/remove_nth_node_from_end.js
/* Given a linked list, remove the nth node from the end of list and return 
 * its head.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    for (var nth = head, i = 0; i < n; ++i)
        nth = nth.next
    if (!nth)
        return head.next
    for (var current = head; nth.next; nth = nth.next)
        current = current.next
    current.next = current.next.next
    return head
};
